File: platforms/electron/src/js/CadController.ts
import type {
  CadWasmModule,
  DrawingInfo,
  LayerInfo,
  ViewportState,
  WasmCadEngine,
} from './types';
import { loadWasmModule } from './WasmLoader';

/**
 * High-level TypeScript facade wrapping the WASM CadEngine.
 *
 * Usage:
 * ```ts
 * const ctrl = new CadController();
 * await ctrl.initialize(canvas);
 * await ctrl.loadFile('/drawings/part.dxf');
 * ctrl.renderFrame();
 * ```
 */
export class CadController {
  private engine: WasmCadEngine | null = null;
  private module: CadWasmModule | null = null;

  // ------------------------------------------------------------------ //
  //  Lifecycle
  // ------------------------------------------------------------------ //

  /**
   * Load the WASM module, create a CadEngine, and initialise it against
   * the given HTML canvas element.
   */
  async initialize(
    canvas: HTMLCanvasElement,
    wasmPath?: string,
  ): Promise<void> {
    this.module = await loadWasmModule(wasmPath);
    this.engine = new this.module.CadEngine();

    const width = canvas.clientWidth || canvas.width;
    const height = canvas.clientHeight || canvas.height;

    const rc = this.engine.initialize(canvas, width, height);
    if (rc !== 0) {
      throw new Error(`CadEngine.initialize() failed with code ${rc}`);
    }
  }

  /** Whether the engine has been initialised and is ready to use. */
  get isReady(): boolean {
    return this.engine !== null;
  }

  // ------------------------------------------------------------------ //
  //  File loading
  // ------------------------------------------------------------------ //

  /** Load a DXF/DWG from a raw byte buffer. */
  loadBuffer(data: Uint8Array): void {
    this.requireEngine();
    this.engine!.loadBuffer(data);
  }

  /** Fetch a remote file and load it into the engine. */
  async loadFile(url: string): Promise<void> {
    this.requireEngine();
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`Failed to fetch "${url}": ${response.status} ${response.statusText}`);
    }
    const arrayBuffer = await response.arrayBuffer();
    this.engine!.loadBuffer(new Uint8Array(arrayBuffer));
  }

  /** Close the current file and free associated resources. */
  closeFile(): void {
    this.requireEngine();
    this.engine!.closeFile();
  }

  // ------------------------------------------------------------------ //
  //  Rendering
  // ------------------------------------------------------------------ //

  /** Render a single frame. */
  renderFrame(): void {
    this.requireEngine();
    this.engine!.renderFrame();
  }

  /** Notify the engine that the viewport size has changed. */
  resize(width: number, height: number): void {
    this.requireEngine();
    this.engine!.resize(width, height);
  }

  // ------------------------------------------------------------------ //
  //  View controls
  // ------------------------------------------------------------------ //

  /** Pan the camera by (dx, dy) in screen-pixel units. */
  pan(dx: number, dy: number): void {
    this.requireEngine();
    this.engine!.pan(dx, dy);
  }

  /**
   * Zoom by `factor` around screen point (pivotX, pivotY).
   * factor > 1 zooms in, factor < 1 zooms out.
   */
  zoom(factor: number, pivotX: number, pivotY: number): void {
    this.requireEngine();
    this.engine!.zoom(factor, pivotX, pivotY);
  }

  /** Fit the drawing to fill the viewport. */
  fitToExtents(): void {
    this.requireEngine();
    this.engine!.fitToExtents();
  }

  // ------------------------------------------------------------------ //
  //  Queries
  // ------------------------------------------------------------------ //

  getDrawingInfo(): DrawingInfo {
    this.requireEngine();
    return this.engine!.getDrawingInfo();
  }

  getViewportState(): ViewportState {
    this.requireEngine();
    return this.engine!.getViewportState();
  }

  getLayers(): LayerInfo[] {
    this.requireEngine();
    return this.engine!.getLayers();
  }

  setLayerVisibility(name: string, visible: boolean): void {
    this.requireEngine();
    this.engine!.setLayerVisibility(name, visible);
  }

  // ------------------------------------------------------------------ //
  //  Cleanup
  // ------------------------------------------------------------------ //

  /** Shutdown the engine and release all WASM-side resources. */
  destroy(): void {
    if (this.engine) {
      this.engine.shutdown();
      this.engine.delete();
      this.engine = null;
    }
    this.module = null;
  }

  // ------------------------------------------------------------------ //
  //  Internal helpers
  // ------------------------------------------------------------------ //

  private requireEngine(): asserts this is { engine: WasmCadEngine } {
    if (!this.engine) {
      throw new Error('CadController is not initialized. Call initialize() first.');
    }
  }
}

File: platforms/flutter/lib/src/cad_controller.dart
import 'dart:ffi';
import 'dart:typed_data';
import 'package:ffi/ffi.dart';
import 'cad_ffi_bindings.dart';

/// High-level Dart wrapper for the CAD engine.
class CadController {
  final CadEngineBindings _bindings;
  Pointer<CadEngine>? _engine;

  CadController() : _bindings = CadEngineBindings();

  bool get isInitialized => _engine != null;

  /// Initialize the engine. Call once before loading files.
  void initialize({int width = 800, int height = 600}) {
    _engine = _bindings.create();
    _bindings.initialize(_engine!, nullptr, width, height);
  }

  /// Load a CAD file by path.
  bool loadFile(String path) {
    if (_engine == null) return false;
    final cPath = path.toNativeUtf8();
    try {
      return _bindings.loadFile(_engine!, cPath) == 0;
    } finally {
      malloc.free(cPath);
    }
  }

  /// Load a CAD file from raw bytes (DXF or DWG).
  bool loadBuffer(Uint8List data) {
    if (_engine == null) return false;
    final buf = malloc<Uint8>(data.length);
    try {
      buf.asTypedList(data.length).setAll(0, data);
      return _bindings.loadBuffer(_engine!, buf, data.length) == 0;
    } finally {
      malloc.free(buf);
    }
  }

  /// Get the render command buffer (for Flutter Canvas rendering).
  Uint8List? getCommandBuffer() {
    if (_engine == null) return null;
    final ptr = _bindings.getCommandBuffer(_engine!);
    final size = _bindings.getCommandBufferSize(_engine!);
    if (ptr == nullptr || size == 0) return null;
    return ptr.asTypedList(size);
  }

  /// Pan the view by (dx, dy) in screen pixels.
  void pan(double dx, double dy) {
    if (_engine == null) return;
    _bindings.pan(_engine!, dx, dy);
  }

  /// Zoom by factor around pivot (px, py) in screen coords.
  void zoom(double factor, double px, double py) {
    if (_engine == null) return;
    _bindings.zoom(_engine!, factor, px, py);
  }

  /// Fit view to drawing extents.
  void fitToExtents() {
    if (_engine == null) return;
    _bindings.fitToExtents(_engine!);
  }

  /// Notify engine of canvas resize.
  void resize(int width, int height) {
    if (_engine == null) return;
    _bindings.resize(_engine!, width, height);
  }

  /// Get drawing extents as (minX, minY, maxX, maxY).
  (double, double, double, double)? getExtents() {
    if (_engine == null) return null;
    final pMinX = malloc<Float>(), pMinY = malloc<Float>();
    final pMaxX = malloc<Float>(), pMaxY = malloc<Float>();
    try {
      _bindings.getExtents(_engine!, pMinX, pMinY, pMaxX, pMaxY);
      return (pMinX.value, pMinY.value, pMaxX.value, pMaxY.value);
    } finally {
      malloc.free(pMinX);
      malloc.free(pMinY);
      malloc.free(pMaxX);
      malloc.free(pMaxY);
    }
  }

  /// Release engine resources.
  void dispose() {
    if (_engine != null) {
      _bindings.shutdown(_engine!);
      _bindings.destroy(_engine!);
      _engine = null;
    }
  }
}
